### PR TITLE
feat: Use LocalScheduler framework to implement FlushScheduler

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
This PR uses the scheduler framework to re-implement FlushScheduler
- Adds FlushRequest and FlushHandler. The FlushHandler still uses the FlushJob to flush memtables
- The key of the FlushRequest is the region id and the flush sequence
- The FlushScheduler uses a LocalScheduler to schedule flush request
- Removes the background mod
- Refactors the RegionMap to a new type to make other structs like the FlushScheduler possible to hold the region map

### Future Works
After this PR is merged, I plan to add a background picker task to the new FlushScheduler and implement flush strategies mentioned in #1462

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
- #1462

### CODE REVIEW

File: src/storage/src/error.rs [hunk 1]
```diff
@@ -138,21 +138,13 @@ pub enum Error {
         source: std::io::Error,
     },
 
-    #[snafu(display("Failed to join task, source: {}", source))]
-    JoinTask {
-        source: common_runtime::JoinError,
+    #[snafu(display("Failed to wait flush, region_id: {}, source: {}", region_id, source))]
```
Comment (killme2008) [inline src/storage/src/error.rs:141]:
```suggestion
    #[snafu(display("Failed to wait flushing, region_id: {}, source: {}", region_id, source))]
```

File: src/storage/src/error.rs [hunk 2]
```diff
@@ -539,9 +540,10 @@ impl ErrorExt for Error {
             ConvertChunk { source, .. } => source.status_code(),
             MarkWalObsolete { source, .. } => source.status_code(),
             DecodeParquetTimeRange { .. } => StatusCode::Unexpected,
-            RateLimited { .. } | StopScheduler { .. } | CompactTaskCancel { .. } => {
-                StatusCode::Internal
-            }
+            RateLimited { .. }
+            | StopScheduler { .. }
+            | CompactTaskCancel { .. }
+            | DuplicateFlush { .. } => StatusCode::Internal,
```
Comment (killme2008) [inline src/storage/src/error.rs:546]: `DuplicateFlush` prefer to return `InvalidArguments`.

File: src/storage/src/flush/scheduler.rs [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+        Ok(())
+    }
+}
+
+async fn execute_flush<S: LogStore>(
+    req: FlushRequest<S>,
+    compaction_scheduler: CompactionSchedulerRef<S>,
+) {
+    let mut flush_job = FlushJob {
```
Comment (killme2008) [inline src/storage/src/flush/scheduler.rs:190]: I'd like to impl  `From<FlushRequest<S>>` for `FlushJob`.

File: src/storage/src/flush/scheduler.rs [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    };
+
+    if let Err(e) = flush_job.run().await {
+        logging::error!(e; "Failed to flush regoin {}", req.region_id());
+
+        req.complete(Err(e));
+    } else {
+        logging::info!("Successfully flush region: {}", req.region_id());
+
+        let compaction_request = CompactionRequestImpl {
```
Comment (killme2008) [inline src/storage/src/flush/scheduler.rs:209]: The same as above suggestion, impl `From` trait.

File: src/storage/src/flush/scheduler.rs [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+        manifest: req.manifest.clone(),
+        engine_config: req.engine_config.clone(),
+    };
+
+    if let Err(e) = flush_job.run().await {
+        logging::error!(e; "Failed to flush regoin {}", req.region_id());
+
+        req.complete(Err(e));
+    } else {
+        logging::info!("Successfully flush region: {}", req.region_id());
```
Comment (killme2008) [inline src/storage/src/flush/scheduler.rs:207]: The log may be annoying. `debug` is enough.

File: src/storage/src/flush/scheduler.rs [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+}
+
+/// Flush scheduler.
+pub struct FlushScheduler<S: LogStore> {
+    scheduler: LocalScheduler<FlushRequest<S>>,
+}
+
+pub type FlushSchedulerRef<S> = Arc<FlushScheduler<S>>;
+
+impl<S: LogStore> FlushScheduler<S> {
```
Comment (killme2008) [inline src/storage/src/flush/scheduler.rs:118]: Suggestion: adds some metrics to scheduler and job.

File: src/storage/src/flush/scheduler.rs (hunk 3 above)
Comment (fengjiachun) [reply src/storage/src/flush/scheduler.rs:190]: `FlushRequest` and `FlushJob` are not equivalent, there is data loss during the conversion process. Is this suitable for `From`? Is `to_flush_job` will be better?

File: src/storage/src/error.rs (hunk 2 above)
Comment (evenyag) [reply src/storage/src/error.rs:546]: This is an unexpected issue, so I use code internal. Maybe `StatusCode::Unexpected` is better.

File: src/storage/src/scheduler.rs [hunk 7]
```diff
@@ -326,22 +339,14 @@ mod tests {
...
+        /// Users should only call this once.
         async fn wait(&self) {
-            let notify = Arc::new(Notify::new());
-            {
-                let notify = notify.clone();
-                let mut notifies = self.notifies.write().unwrap();
-                notifies.push(notify);
-            }
-            notify.notified().await
+            self.notify.notified().await
```
[the remark this answers is not in the document]
Comment (evenyag) [reply src/storage/src/scheduler.rs:349]: tokio's Notify doesn't have spurious wakeup so I think we don't need this loop.

File: src/storage/src/flush/scheduler.rs (hunk 3 above)
Comment (evenyag) [reply src/storage/src/flush/scheduler.rs]: `From<FlushRequest<S>>` consumes the request. I use `From<&FlushRequest<S>>` instead.

**Review by killme2008 — APPROVED** [review]
A little suggestion. LGTM

**Pull request merged**